internal/api/courses/assignments/submissions/proxy/resubmit.go

```diff
@@ -0,0 +1,113 @@
+}
+
+// Proxy resubmit an assignment submission to the autograder.
+func HandleResubmit(request *ResubmitRequest) (*ResubmitResponse, *core.APIError) {
+	response := ResubmitResponse{}
+
+	if !request.ProxyUser.Found {
+		return &response, nil
+	}
+
+	response.FoundUser = true
+
+	gradingResult, err := db.GetSubmissionContents(request.Assignment, request.ProxyUser.Email, request.TargetSubmission)
+	if err != nil {
+		return nil, core.NewInternalError("-632", &request.APIRequestCourseUserContext, "Failed to get submission contents.").
+			Err(err).Assignment(request.Assignment.GetID()).
+			Add("target-user", request.ProxyUser.Email).Add("submission", request.TargetSubmission)
+	}
+
+	if gradingResult == nil {
+		return &response, nil
+	}
+
+	response.FoundSubmission = true
+
+	tempDir, err := util.MkDirTemp("resumbit-request-files-")
+	if err != nil {
+		return nil, core.NewInternalError("-633", &request.APIRequestCourseUserContext, "Failed to create temp resubmit files directory.").
+			Err(err).Assignment(request.Assignment.GetID()).
+			Add("target-user", request.ProxyUser.Email).Add("submission", request.TargetSubmission)
+	}
+
+	err = util.GzipBytesToDirectory(tempDir, gradingResult.InputFilesGZip)
+	if err != nil {
+		util.RemoveDirent(tempDir)
+		return nil, core.NewInternalError("-634", &request.APIRequestCourseUserContext, "Failed to write submission input to a temp dir.").
+			Err(err).Assignment(request.Assignment.GetID()).
+			Add("target-user", request.ProxyUser.Email).Add("submission", request.TargetSubmission)
+	}
+
+	gradeOptions := grader.GetDefaultGradeOptions()
+	// Proxy submissions are not subject to submission restrictions.
+	gradeOptions.CheckRejection = false
+	gradeOptions.ProxyUser = request.User.Email
+	gradeOptions.ProxyTime = grader.ResolveProxyTime(request.ProxyTime, request.Assignment)
+
+	result, reject, failureMessage, err := grader.Grade(request.Context, request.Assignment, tempDir, request.ProxyUser.Email, request.Message, gradeOptions)
+	if err != nil {
+		stdout := ""
+		stderr := ""
+
+		if (result != nil) && (result.HasTextOutput()) {
+			stdout = result.Stdout
+			stderr = result.Stderr
+		}
+
+		log.Warn("Resubmission failed internally.", err, request.Assignment, log.NewAttr("stdout", stdout), log.NewAttr("stderr", stderr), request.User)
+
+		return &response, nil
+	}
+
+	// A proxy resubmission should never be rejected.
+	if reject != nil {
+		log.Error("Proxy resubmission rejected.", request.Assignment, log.NewAttr("reason", reject.String()), log.NewAttr("request", request), request.User)
+
+		response.Rejected = true
+		response.Message = reject.String()
+		return &response, nil
+	}
+
+	if failureMessage != "" {
+		log.Debug("Resubmission got a soft error.", request.Assignment, log.NewAttr("message", failureMessage), log.NewAttr("request", request), request.User)
+
+		response.Message = failureMessage
+		return &response, nil
+	}
+
+	response.GradingSuccess = true
+	response.GradingInfo = result.Info
```

Comment:
This is the third place we have code similar to this.
See if it can be easily combined.
(Note that I am not looking into the other endpoints right now, so there could be some issue that makes them hard to combine.)

It may also be possible to combine most of the outputs (response fields) into a single struct that can be embedded and then used for this common function.

It would also not be a problem if all the log messages look similar (because the request will get logged with a unique ID).